src/UsgsAstroSarSensorModel.cpp

```diff
@@ -136,3 +141,209 @@ string UsgsAstroSarSensorModel::constructStateFromIsd(
+    }
+  }
+  return csm::ImageCoord(-1, -1); // what do we do if this fails?
+}
+
+// Calculate the root 
+double UsgsAstroSarSensorModel::dopplerShiftRoot(
+    csm::EcefCoord groundPt) const{
+
+  // Moon body-fixed coordinates of surface point
+  csm::EcefCoord surfPt(0.0, 0.0, 0.0); // obviously populate with real value
+
+   std::function<double(double)> dopplerShiftFrequency = [surfPt](double time) { 
+     // Exsiting vector subtraction function in usgscsm? 
+     double spacecraftPosition[3] = {3.73740000e+06,  0.00000000e+00, -0.00000000e+00}; //spacecraftPosition(time);
+     double spacecraftVelocity[3] = {-0.00000000e+00,  0.00000000e+00, -3.73740000e+0};; //spacecraftPosition(time);
+     double lookVector[3];
+
+     lookVector[0] = spacecraftPosition[0] - surfPt.x;
+     lookVector[1] = spacecraftPosition[1] - surfPt.y;
+     lookVector[2] = spacecraftPosition[2] - surfPt.z;
+
+     // Existing vector magnitude function in usgscsm? 
+     double slantRange = sqrt(pow(lookVector[0], 2) +  pow(lookVector[1], 2) + pow(lookVector[2], 2)); 
+
+
+     // Existing dot product function in usgscsm?
+     double dopplerShift = -2.0 * (lookVector[0]*spacecraftVelocity[0] + lookVector[1]*spacecraftVelocity[1] 
+                            + lookVector[2]*spacecraftVelocity[2])/(slantRange * WAVELENGTH);
+     return dopplerShift;
+   };
+
+  // Lower-bound for doppler-shift
+  double startShift = dopplerShiftFrequency(START_TIME);
+
+  // Upper-bound for doppler-shift
+  double stopShift = dopplerShiftFrequency(STOP_TIME);
+
+  // Make sure we bound root (dopplerShift = 0.0)
+  if (((startShift < 0.0) && (stopShift < 0.0)) || ((startShift > 0.0) && (stopShift > 0.0))) {
+    return -1;
+  }
+
+  // Do root-finding for "dopplerShift"
+  double tolerance = (STOP_TIME - START_TIME) / LINES / 20.0;
+  double maxIterations = 30;
+
+  // Order the bounds
+  double fl, fh, xl, xh;
+  if (startShift < stopShift) {
+    xl = START_TIME;
+    xh = STOP_TIME;
+  }
+  else {
+    xl = STOP_TIME;
+    xh = START_TIME;
+  }
+
+  return secantRoot(dopplerShiftFrequency, xl, xh, tolerance, 30);
+}
+
+
+double secantRoot(std::function<double(double)> lambda, double xl, double xh, 
+                  double tolerance, double maxIters) {
+  double time;
+  int j = 0;
+    int found = -1;
+    double shift;
+    double fl = lambda(xl);
+    double fh = lambda(xh);
+    while (found==-1 && j < maxIters) {
+      // Use the secant method to guess the next time
+      time = xl + (xh - xl) * fl / (fl - fh);
+      
+      // Compute a guessed Doppler shift.
+      shift = lambda(time);
+      
+      // Update the bounds
+      double deltaTime;
+      if (shift < 0.0) {
+        deltaTime = xl - time;
+        xl = time;
+        fl = shift;
+      }
+      else {
+        deltaTime = xh - time;
+        xh = time;
+        fh = shift;
+      }
+
+      // See if we are done
+      if ((fabs(deltaTime) <= tolerance) || (shift == 0.0)) {
+        time = time;
```

Comment:
time = time?

Reply:
Cleaning this up, not sure what that was supposed to be